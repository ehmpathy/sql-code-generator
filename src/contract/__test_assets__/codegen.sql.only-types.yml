language: mysql
dialect: 5.7
resources:
  - 'schema/**/*.sql'
queries:
  - 'src/dao/**/*.ts'
  - '!src/**/*.test.ts'
  - '!src/**/*.test.integration.ts'
generates:
  types: src/generated/fromSql/types.ts
