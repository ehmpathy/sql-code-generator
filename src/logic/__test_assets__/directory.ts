export const TEST_ASSETS_ROOT_DIR = __dirname;
