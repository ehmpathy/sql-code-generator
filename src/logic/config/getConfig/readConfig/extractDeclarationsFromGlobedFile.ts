import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import strip from 'sql-strip-comments';

import { QueryDeclaration, ResourceDeclaration } from '../../../../domain';
import { extractSqlFromFile } from '../../../common/extractSqlFromFile';

export enum DeclarationType {
  QUERY = 'QUERY',
  RESOURCE = 'RESOURCE",',
}

export const extractDeclarationsFromGlobedFile = async ({
  rootDir,
  relativePath,
  type,
}: {
  rootDir: string;
  relativePath: string;
  type: DeclarationType;
}) => {
  // define the file path
  const filePath = `${rootDir}/${relativePath}`;

  // extract the sql
  const sql = await extractSqlFromFile({ filePath });

  // define the declaration for a query
  if (type === DeclarationType.QUERY) {
    // note: we only support one declaration per query, since we require the comments to define the query name
    return [
      new QueryDeclaration({
        path: relativePath,
        sql,
      }),
    ];
  }

  // define the declarations for a resource
  if (type === DeclarationType.RESOURCE) {
    // split up the sql by create resource statements, to support multiple resources per sql file
    const sqlCreateResourceStatements = sql
      .split(/(create(?:\sor\sreplace)?\s(?:table|function|view))/gi)
      .reduce((pairs, thisPart) => {
        // define what kind of part this is
        const isThisPartCreateDeclaration = thisPart
          .trim()
          .toLowerCase()
          .startsWith('create');

        // define the last pair
        const lastPair:
          | { createDeclaration: string; resourceDeclaration?: string }
          | undefined = pairs.slice(-1)[0];

        // if the last pair is not defined yet
        if (!lastPair) {
          // if this part is not a create declaration, then do nothing; we can ignore the prefixes
          if (!isThisPartCreateDeclaration) return pairs;

          // otherwise, since it is a create declaration, start the first pair
          return [{ createDeclaration: thisPart }];
        }

        // if the last pair already has a resource declaration
        if (lastPair.resourceDeclaration) {
          // if this part is a create declaration, then start a new pair
          if (isThisPartCreateDeclaration)
            return [...pairs, { createDeclaration: thisPart }];

          // otherwise, this is unexpected, so fail fast
          throw new UnexpectedCodePathError(
            'lastPair.resourceDeclaration already exists but thisPart.!isCreateDeclaration. how is that possible?',
            { lastPair, thisPart },
          );
        }

        // if the last pair does not have a resource declaration
        if (!lastPair.resourceDeclaration) {
          // if this part is a resource declaration, then complete this pair
          if (!isThisPartCreateDeclaration)
            return [
              ...pairs.slice(0, -1),
              { ...lastPair, resourceDeclaration: thisPart },
            ];

          // otherwise, this is unexpected, so fail fast
          throw new UnexpectedCodePathError(
            'lastPair.resourceDeclaration does not exist, but thisPart.isCreateDeclaration. how is that possible?',
            {
              lastPair,
              thisPart,
            },
          );
        }

        // we should never reach here, as the above should have captured every scenario, so fail fast
        throw new UnexpectedCodePathError('should never reach here. why?', {
          lastPair,
          thisPart,
        });
      }, [] as { createDeclaration: string; resourceDeclaration?: string }[])
      .map((pair) =>
        [pair.createDeclaration, pair.resourceDeclaration].join(''),
      );

    // return a declaration for each
    return sqlCreateResourceStatements.map(
      (sql) =>
        new ResourceDeclaration({
          path: relativePath,
          sql,
        }),
    );
  }
  throw new Error('unexpected'); // fail fast
};
