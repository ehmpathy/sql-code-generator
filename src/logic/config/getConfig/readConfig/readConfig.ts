import { GeneratorConfig } from '../../../../domain';
import { getAllPathsMatchingGlobs } from '../getAllPathsMatchingGlobs/getAllPathsMatchingGlobs';
import {
  DeclarationType,
  extractDeclarationFromGlobedFile,
} from './extractDeclarationFromGlobedFile';
import { readYmlFile } from './utils/readYmlFile';

/*
  1. read the yml file
  2. get the connection (i.e., import from the specified connection file)
  3. get the definitions and flatten them
*/
export const readConfig = async ({ filePath }: { filePath: string }) => {
  const configDir = filePath.split('/').slice(0, -1).join('/'); // drops the file name

  // get the yml
  const contents = await readYmlFile({ filePath });

  // get the output dir
  if (!contents.generates) throw new Error('generates key must be defined');
  if (!contents.generates.types)
    throw new Error(
      'generates.types must specify where to output the generated types',
    );

  // get the language and dialect
  if (!contents.language) throw new Error('language must be defined');
  const language = contents.language;
  if (!contents.dialect) throw new Error('dialect must be defined');
  const dialect = `${contents.dialect}`; // ensure that we read it as a string, as it could be a number

  // get the resource declarations
  const resourceGlobs = contents.resources;
  const resourcePaths = await getAllPathsMatchingGlobs({
    globs: resourceGlobs,
    root: configDir,
  });
  const resourceDeclarations = await Promise.all(
    resourcePaths
      .sort() // for determinism in order
      .map((relativePath) =>
        extractDeclarationFromGlobedFile({
          rootDir: configDir,
          relativePath,
          type: DeclarationType.RESOURCE,
        }),
      ),
  );

  // get the query declarations
  const queryDeclarations = await (async () => {
    const queryGlobs = contents.queries;
    if (!queryGlobs) return [];
    const queryPaths = await getAllPathsMatchingGlobs({
      globs: queryGlobs,
      root: configDir,
    });
    return await Promise.all(
      queryPaths
        .sort() // for determinism in order
        .map((relativePath) =>
          extractDeclarationFromGlobedFile({
            rootDir: configDir,
            relativePath,
            type: DeclarationType.QUERY,
          }),
        ),
    );
  })();

  // return the results
  return new GeneratorConfig({
    rootDir: configDir,
    generates: {
      types: contents.generates.types,
      queryFunctions: contents.generates.queryFunctions,
    },
    language,
    dialect,
    declarations: [...resourceDeclarations, ...queryDeclarations],
  });
};
