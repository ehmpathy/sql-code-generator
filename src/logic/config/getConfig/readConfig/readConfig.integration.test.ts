import { GeneratorConfig } from '../../../../domain';
import { TEST_ASSETS_ROOT_DIR } from '../../../__test_assets__/directory';
import { readConfig } from './readConfig';

describe('readConfig', () => {
  it('should be able to read a fully declared config', async () => {
    const config = await readConfig({
      filePath: `${TEST_ASSETS_ROOT_DIR}/exampleProject/mysql/codegen.sql.yml`,
    });
    expect(config).toBeInstanceOf(GeneratorConfig);
    expect(config.language).toEqual('mysql');
    expect(config.dialect).toEqual('5.7');
    expect(config.declarations.length).toEqual(12);
    expect({ ...config, rootDir: '__DIR__' }).toMatchSnapshot(); // to log an example of the output; note we mask dir to make it machine independent
  });
  it('should be able to read a config without queries', async () => {
    const config = await readConfig({
      filePath: `${TEST_ASSETS_ROOT_DIR}/exampleProject/postgres-noqueries/codegen.sql.yml`,
    });
    expect(config).toBeInstanceOf(GeneratorConfig);
    expect(config.language).toEqual('postgres');
    expect(config.dialect).toEqual('10.7');
    expect(config.declarations.length).toEqual(3);
    expect({ ...config, rootDir: '__DIR__' }).toMatchSnapshot(); // to log an example of the output; note we mask dir to make it machine independent
  });
});
