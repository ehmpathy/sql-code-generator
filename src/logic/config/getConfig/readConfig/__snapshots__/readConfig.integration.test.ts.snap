// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readConfig should be able to read a config without queries 1`] = `
{
  "declarations": [
    ResourceDeclaration {
      "path": "schema/functions/get_answer_to_life.sql",
      "sql": "CREATE OR REPLACE FUNCTION get_answer_to_life()
RETURNS int
LANGUAGE plpgsql
AS $$
  BEGIN
    RETURN 42;
  END;
$$
",
    },
    ResourceDeclaration {
      "path": "schema/tables/ice_cream.sql",
      "sql": "CREATE TABLE \`ice_cream\` (
  \`id\` bigint(20) NOT NULL AUTO_INCREMENT,
  \`uuid\` char(36) COLLATE utf8mb4_bin NOT NULL,
  \`created_at\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  \`name\` varchar(255) COLLATE utf8mb4_bin NOT NULL,
  \`ingredient_ids_hash\` binary(32) NOT NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE KEY \`tag_ux1\` (\`name\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
",
    },
    ResourceDeclaration {
      "path": "schema/tables/ingredient.sql",
      "sql": "CREATE TABLE ingredient (
  id bigserial PRIMARY KEY,
  uuid uuid NOT NULL,
  created_at timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  name varchar NOT NULL,
  cost numeric(5, 2) NOT NULL,
  CONSTRAINT ingredient_pk PRIMARY KEY (id),
  CONSTRAINT ingredient_ux1 UNIQUE (name)
)
",
    },
  ],
  "dialect": "10.7",
  "generates": {
    "queryFunctions": undefined,
    "types": "src/generated/fromSql/types.ts",
  },
  "language": "postgres",
  "rootDir": "__DIR__",
}
`;

exports[`readConfig should be able to read a fully declared config 1`] = `
{
  "declarations": [
    ResourceDeclaration {
      "path": "schema/functions/hash_string.sql",
      "sql": "CREATE FUNCTION \`get_from_delimiter_split_string\`(
  in_array varchar(255),
  in_delimiter char(1),
  in_index int
) RETURNS varchar(255) CHARSET utf8mb4 COLLATE utf8mb4_unicode_ci
BEGIN
  /*
    usage:
      SELECT get_from_delimiter_split_string('1,5,3,7,4', ',',  1); -- returns '5'
      SELECT get_from_delimiter_split_string('1,5,3,7,4', ',',  10); -- returns ''
  */
  return REPLACE( -- remove the delimiters after doing the following:
    SUBSTRING( -- pick the string
      SUBSTRING_INDEX(in_array, in_delimiter, in_index + 1), -- from the string up to index+1 counts of the delimiter
      LENGTH(
        SUBSTRING_INDEX(in_array, in_delimiter, in_index) -- keeping only everything after index counts of the delimiter
      ) + 1
    ),
    in_delimiter,
    ''
  );
END
",
    },
    ResourceDeclaration {
      "path": "schema/functions/upsert_image.sql",
      "sql": "CREATE FUNCTION \`upsert_image\`(
  in_url varchar(190),
  in_caption varchar(190),
  in_credit varchar(190),
  in_alt_text varchar(190)
) RETURNS bigint(20)
BEGIN
  -- declarations
  DECLARE v_static_id BIGINT;
  DECLARE v_created_at DATETIME(6) DEFAULT CURRENT_TIMESTAMP(6); -- define a common created_at timestamp to use

  -- find or create the static entity
  SET v_static_id = (
    SELECT id
    FROM image
    WHERE 1=1
      AND (url = BINARY in_url)
      AND (caption = BINARY in_caption OR (caption IS null AND in_caption IS null))
      AND (credit = BINARY in_credit OR (credit IS null AND in_credit IS null))
      AND (alt_text = BINARY in_alt_text OR (alt_text IS null AND in_alt_text IS null))
  );
  IF (v_static_id IS NULL) THEN -- if entity could not be found originally, create the static entity
    INSERT INTO image
      (uuid, created_at, url, caption, credit, alt_text)
      VALUES
      (uuid(), v_created_at, in_url, in_caption, in_credit, in_alt_text);
    SET v_static_id = (
      SELECT last_insert_id()
    );
  END IF;

  -- return the static entity id
  return v_static_id;
END
",
    },
    ResourceDeclaration {
      "path": "schema/tables/communication_channel.sql",
      "sql": "CREATE TABLE \`communication_channel\` (
  -- meta
  \`id\` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
  \`uuid\` VARCHAR(36) NOT NULL,
  \`created_at\` DATETIME DEFAULT CURRENT_TIMESTAMP,

  -- static data
  \`type\` ENUM('APP', 'SMS', 'EMAIL') NOT NULL, -- e.g., "how do we send this"
  \`address\` VARCHAR(36) NOT NULL, -- e.g., "where do we send this"

  -- meta meta
  UNIQUE (\`type\`, \`address\`) -- type + address uniquely identify a communication channel
) ENGINE = InnoDB;
",
    },
    ResourceDeclaration {
      "path": "schema/tables/ice_cream.sql",
      "sql": "CREATE TABLE \`ice_cream\` (
  \`id\` bigint(20) NOT NULL AUTO_INCREMENT,
  \`uuid\` char(36) COLLATE utf8mb4_bin NOT NULL,
  \`created_at\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  \`name\` varchar(255) COLLATE utf8mb4_bin NOT NULL,
  \`ingredient_ids_hash\` binary(32) NOT NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE KEY \`tag_ux1\` (\`name\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
",
    },
    ResourceDeclaration {
      "path": "schema/tables/image.sql",
      "sql": "CREATE TABLE \`image\` (
  \`id\` bigint(20) NOT NULL AUTO_INCREMENT,
  \`uuid\` char(36) COLLATE utf8mb4_bin NOT NULL,
  \`created_at\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  \`url\` varchar(190) COLLATE utf8mb4_bin NOT NULL,
  \`caption\` varchar(190) COLLATE utf8mb4_bin DEFAULT NULL,
  \`credit\` varchar(190) COLLATE utf8mb4_bin, -- note: we strip "DEFAULT NULL" here for testing
  \`alt_text\` varchar(190) COLLATE utf8mb4_bin,
  PRIMARY KEY (\`id\`),
  UNIQUE KEY \`image_ux1\` (\`url\`,\`caption\`,\`credit\`,\`alt_text\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
",
    },
    ResourceDeclaration {
      "path": "schema/tables/ingredient.sql",
      "sql": "CREATE TABLE \`ingredient\` (
  \`id\` bigint(20) NOT NULL AUTO_INCREMENT,
  \`uuid\` char(36) COLLATE utf8mb4_bin NOT NULL,
  \`created_at\` datetime(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  \`name\` varchar(255) COLLATE utf8mb4_bin NOT NULL,
  \`cost\` decimal(5,2) NOT NULL,
  PRIMARY KEY (\`id\`),
  UNIQUE KEY \`tag_ux1\` (\`name\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
",
    },
    ResourceDeclaration {
      "path": "schema/tables/user.sql",
      "sql": "CREATE TABLE \`user\` (
  -- meta
  \`id\` BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
  \`uuid\` VARCHAR(36) NOT NULL,
  \`created_at\` DATETIME DEFAULT CURRENT_TIMESTAMP,

  -- static data (in this example... in real life this should not be static)
  \`first_name\` VARCHAR(255) NOT NULL,
  \`last_name\` VARCHAR(255) NOT NULL,

  -- meta meta
  UNIQUE (\`first_name\`, \`last_name\`) -- (in this example... in real life, users should not actually be unique on their name)
) ENGINE = InnoDB;
",
    },
    ResourceDeclaration {
      "path": "schema/views/view_user_name.sql",
      "sql": "CREATE VIEW \`view_user_name\` AS
  SELECT
    u.id,
    u.first_name,
    u.last_name
  FROM user u;
",
    },
    QueryDeclaration {
      "path": "src/dao/iceCream/findAllNew.ts",
      "sql": "
-- query_name = find_all_new_ice_cream
SELECT
  s.id,
  s.uuid,
  s.name,
  (
    SELECT GROUP_CONCAT(ice_cream_to_ingredient.ingredient_id ORDER BY ice_cream_to_ingredient.ingredient_id)
    FROM ice_cream_to_ingredient WHERE ice_cream_to_ingredient.ice_cream_id = s.id
  ) as ingredient_ids,
  s.created_at
FROM ice_cream s
WHERE s.created_at > DATE_SUB(NOW(), INTERVAL 7 day) -- new if less than 7 days old
LIMIT :limit
;
",
    },
    QueryDeclaration {
      "path": "src/dao/iceCream/findById.ts",
      "sql": "
-- query_name = find_ice_cream_by_id
SELECT
  s.id,
  s.uuid,
  s.name,
  (
    SELECT GROUP_CONCAT(ice_cream_to_ingredient.ingredient_id ORDER BY ice_cream_to_ingredient.ingredient_id)
    FROM ice_cream_to_ingredient WHERE ice_cream_to_ingredient.ice_cream_id = s.id
  ) as ingredient_ids,
  s.created_at
FROM ice_cream s
;
",
    },
    QueryDeclaration {
      "path": "src/dao/user/findAllByName.ts",
      "sql": "
-- query_name = find_all_by_name
SELECT
  u.id,
  u.first_name
  -- concat(u.first_name, u.last_name) as full_name; TODO: support mysql native functions https://github.com/uladkasach/sql-code-generator/issues/3
FROM user u
WHERE u.first_name = :firstName
",
    },
    QueryDeclaration {
      "path": "src/others/queryUserNameView.ts",
      "sql": "
-- query_name = select_user_name_from_view
SELECT
  u.first_name,
  u.last_name
FROM view_user_name u
WHERE u.id = :userId
",
    },
  ],
  "dialect": "5.7",
  "generates": {
    "queryFunctions": "src/generated/fromSql/queryFunctions.ts",
    "types": "src/generated/fromSql/types.ts",
  },
  "language": "mysql",
  "rootDir": "__DIR__",
}
`;
